code/game/objects/items/weapons/weapon.dm

```diff
@@ -10,3 +9,178 @@
+	charger = owner
+	charge_on()
+datum/action/bayonet_charge/proc/charge_on()
+	if(incooldown)
+		to_chat(charger, "<span class='warning'>Charge is still on cooldown!</span>")
+		return
+	chargewhenmove = TRUE
+	src.add_selected_frame()
+	charger = owner //just in case you didnt activate the action, charger would be null if you do.
+	to_chat(charger, "<span class='warning'>You stand ready to charge</span>")
+	RegisterSignal(charger, COMSIG_MOVABLE_MOVED, .proc/update_charging)
+	RegisterSignal(charger, COMSIG_ATOM_DIR_CHANGE, .proc/on_dir_change)
+
+datum/action/bayonet_charge/proc/charge_off()
+	chargewhenmove = FALSE
+	charger = owner
+	to_chat(charger, "<span class='warning'>You no longer stand ready to charge</span>")
+	if(charger) //locks up if you dont check if charger is existant
+		UnregisterSignal(owner, COMSIG_MOVABLE_MOVED)
+		UnregisterSignal(owner, COMSIG_ATOM_DIR_CHANGE)
+		UnregisterSignal(owner, list(COMSIG_MOVABLE_PREBUMP_TURF, COMSIG_MOVABLE_PREBUMP_MOVABLE, COMSIG_MOVABLE_PREBUMP_EXIT_MOVABLE))
+	src.remove_selected_frame()
+	stop_charge()
+
+
+datum/action/bayonet_charge/proc/on_dir_change(datum/source, old_dir, new_dir)
+	SIGNAL_HANDLER
+	if(old_dir == new_dir)
+		return
+	stop_charge()
+//called when the marine moves
+datum/action/bayonet_charge/proc/update_charging(datum/source, atom/oldloc, direction, Forced)
+	SIGNAL_HANDLER_DOES_SLEEP
+	if(movementcontinuationlimit && world.time > movementcontinuationlimit) //if you have stopped moving, it resetss the steps taken coutner
+		stop_charge()
+		movementcontinuationlimit = null
+		return
+	stepstaken++
+	movementcontinuationlimit = world.time + 10
+
+	handle_charge()
+
+datum/action/bayonet_charge/proc/handle_charge()
+
+	//actual charging is under here
+	var/chargespeed = clamp(stepstaken * speedperstep, 0, maxspeed)
+	if(charger.incapacitated())
+		return
+
+	if(charger.pulling)
+		return
+	if(charger.getStaminaLoss() >= 0)
+		charge_off()
+		return
+	if(stepstaken < minactivationrundist)
+		return
+
+	//stuff in this if only happens once the charger starts charging, and not anytime else.
+	if(!amcharging)
+		charger.emote("warcry")
+		RegisterSignal(charger, list(COMSIG_MOVABLE_PREBUMP_TURF, COMSIG_MOVABLE_PREBUMP_MOVABLE, COMSIG_MOVABLE_PREBUMP_EXIT_MOVABLE), .proc/on_contact_with_enemy, TRUE)
+	amcharging = TRUE
+	charger.add_movespeed_modifier(MOVESPEED_ID_MARINE_CHARGE, TRUE, 100, NONE, TRUE, -chargespeed)
+	charger.toggle_move_intent(MOVE_INTENT_WALK) //if this aint here, the sprint compounds with the speed increase of the modifier, making you faster than a fucking charger
+	charger.adjustStaminaLoss(stamlossmult * chargespeed)
+
+
+
+datum/action/bayonet_charge/proc/stop_charge()
+	amcharging = FALSE
+	if(charger)
+		charger.remove_movespeed_modifier(MOVESPEED_ID_MARINE_CHARGE)
+	stepstaken = 0
+
+datum/action/bayonet_charge/proc/on_contact_with_enemy(datum/source, atom/crushed)
+	if(istype(crushed, /mob/living) && amcharging == TRUE)
+		if(isliving(crushed))
```

Comment:
isliving and istype(crushed, /mob/living) is exactly the samething.

With components, you will be able to scrap amcharging var